File: myeong_da_yeon/week1/섬연결하기.java
import java.util.*;

class Solution {
    
    static int[] unf;
    
    public int solution(int n, int[][] costs) {
        int answer = 0;
        
        ArrayList<Point> p = new ArrayList<>();
        
        unf = new int[n];
        
        for (int i = 0 ; i< costs.length;i++) {
            p.add(new Point(costs[i][0], costs[i][1], costs[i][2]));     
        }
        
        for(int i = 0;i<n;i++) {
            unf[i] = i;
        } 
        
        
         p.sort((p1,p2) -> p1.c - p2.c);      
        
        for(int i = 0 ;i<p.size();i++) {
            int x = find(p.get(i).x);
            int y = find(p.get(i).y);
            int c = p.get(i).c;
            
            if (x != y) {
                union(x,y);
                answer += c;
            }
            
        }
        
        
        return answer;
    }
    
    static int find(int v) {
        if(v == unf[v]) {
            return v;
        }
        
        return unf[v] = find(unf[v]);
    }
    
    static void union(int s,int e) {
        int x = find(s);
        int y = find(e);
        
        if(x != y) {
            unf[x] = y;
        }
    }
    
    
}

class Point {
    int x;
    int y;
    int c;
    
    public Point(int x,int y,int c) {
        this.x = x;
        this.y = y;
        this.c = c;
    }
    
}

File: shin_jisoo/Greedy/섬연결하기.cpp
#include <string>
#include <vector>
#include <algorithm>

#define max_int 101

using namespace std;

int parents[max_int];

int find(int node) {
    if (node == parents[node]) return node;

    else return parents[node] = find(parents[node]);
}

bool cmp(const vector<int>& a, const vector<int>& b) {
    return a[2] < b[2];
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;

    for (int i = 0;i < n;i++) {
        parents[i] = i;
    }

    sort(costs.begin(), costs.end(), cmp);

    for (int i = 0;i < costs.size();i++) {
        int start = find(costs[i][0]);
        int end = find(costs[i][1]);
        int cost = costs[i][2];

        if (start != end) {
            parents[start] = end;

            answer += cost;
        }
    }
    return answer;
}